### DISCUSSION TITLE
Timeout not busy for too long 10 seconds.

### DISCUSSION THREAD

**Discussion by JoannaSieranskiPS** [ricaun-io/RevitTest#23] [Q&A] [answered]
I'm running a test where I query for ViewSchedules and then read all the cell from these schedules. I then want to check some of the cell text. However, at some point all the test stop with a failure:
```
Details are not available
RevitTest: Timeout not busy for too long 10 seconds.
```

This is even propagated to tests that were already marked as success before.
At first, I thought that this issue was connected to the timeout, but when I run the test samples, the timeout message is different. 
```
Details are not available
RevitTest: Timeout 1 minutes.
```

Nevertheless, I wasn't able to change the timeout too. It seems that the below entry in my code is ignored. 
```
[assembly: System.Reflection.AssemblyMetadata("NUnit.Timeout", "0.01")]
[assembly: System.Reflection.AssemblyMetadata("NUnit.Verbosity", "2")]

public class TestQuery : TestBase
{
        // test code is in here
}
```

What am I missing on the timeout configuration? And how can I prevent the first mentioned error from occurring, so my test can continue?

**Comment by ricaun** [comment]
Hello,

This timeout should not happen. I wonder what code running to trigger that.
```
RevitTest: Timeout not busy for too long 10 seconds.
```
That's hard coded with 10 seconds, should you share some code that trigger that? Don't know how many cell are you reading.

--- 

The `NUnit.Timeout` only affect the whole test timeout, just to prevent is some test is running for too long or in a infinite loop. 
To abort everything, you should not put `0.01`, usually a bigger value like `30` minutes. The default is `10` minutes if you don't add the attribute.

**Reply by ricaun** [reply]
What is this `Details are not available`?

**Reply by paulbuechner** [reply]
The `Details are not available` is a Rider thing, I believe.

Regarding the timeout. Could you please elaborate on what this one is used for and why it shouldn't be triggered in regular circumstances?

I've rerun our tests (also in VS), and the timeout seems to occur if I run multiple tests at once, which then (may) "take too long" to return the "wrapping test runner". At least that's my guess, because when I run tests one by one, or maybe only 2 to 4 at the time, everything works just fine, and no timeout is triggered.

Would it be possible for you to configure the timeout?

**Reply by ricaun** [reply, marked as answer]
@paulbuechner The main reason the `NUnit.Timeout` exist is to prevent infinite loop or some UI that blocks the Revit process and the test never finish.
So if your tests runs for more than 10 minutes (Including Revit opening) the RevitTest would start aborting the tests.
You can change the timeout using the metadata `NUnit.Timeout`:
```c#
[assembly: System.Reflection.AssemblyMetadata("NUnit.Timeout", "30")] // 30 minutes
```
This would change the timeout to 30 minutes, and in total if all the tests could take more than 30 minutes to finish you should put a bigger value.

**Reply by paulbuechner** [reply]
Sure, but

```cs
[assembly: System.Reflection.AssemblyMetadata("NUnit.Timeout", "30")] // 30 minutes
```

is not affecting `RevitTest: Timeout not busy for too long 10 seconds.`

**Reply by ricaun** [reply]
@paulbuechner, sorry the `RevitTest: Timeout not busy for too long 10 seconds.` is a different fail safe not related to the `NUnit.Timeout`.

I wonder what test are you running the make that to trigger. Because that should never trigger in a normal run, is like Revit sending the message to RevitTest that is not busy (only happen when the test finish), and the RevitTest still running waiting for the test data result.

I added the `RICAUN_REVITTEST_CONSOLE_TIMEOUT_NOT_BUSY_MAX_SECONDS` environment variable that you can add in your system to overwrite the 10 seconds, but I don't think gonna fix the issue.

Is more likely Revit is sending the final test result message to RevitTest and the message never shows in the RevitTest process for some reason.

If that happens, does not matter the timeout value, the message is only send once.

What kinda of test are you running? The `RevitTest: Timeout not busy for too long 10 seconds.` always trigger or is something random?

**Reply by paulbuechner** [reply]
Please refer to my previous answer:

> At least that's my guess, because when I run tests one by one, or maybe only 2 to 4 at the time, everything works just fine, and no timeout is triggered.

I can reliably reproduce the "issue", but only if I run the whole batch of tests, making me believe that the timeout is triggered by "tests are taking too long to finish".

I'll try your added env variable and keep you updated. Thank you!

**Reply by paulbuechner** [reply]
When will the new version be available for testing?

**Reply by ricaun** [reply]
> When will the new version be available for testing?

The enviroment variable is already availabe in version 1.8.0.
https://github.com/ricaun-io/ricaun.RevitTest/blob/master/CHANGELOG.md

**Reply by ricaun** [reply]
> I can reliably reproduce the "issue", but only if I run the whole batch of tests, making me believe that the timeout is triggered by "tests are taking too long to finish".
> 

How many tests are you running? Do you have some test with asyn Task? Are you using Console.WriteLine or similar in the tests?

My guess is the final message is failing for some reason.

**Reply by paulbuechner** [reply]
Unfortunately, raising the timeout didn't help much either. It seems like all tests are "done" and then it waits... and after the `RICAUN_REVITTEST_CONSOLE_TIMEOUT_NOT_BUSY_MAX_SECONDS`, the timeout triggers.

> How many tests are you running?

Overall, it takes maybe 20 seconds to run. So not that big of a `TestFixture`...

> Do you have some test with asyn Task?

No.

> Are you using Console.WriteLine?

No. Using `log4net`.

> My guess is the final message is failing for some reason.

What do u mean by that?

**Reply by ricaun** [reply]
> What do u mean by that?

My assumption was the communication was failing, but after some tests.  

I guess the problem is how many tests are running, and the log inside Revit that take some time to execute.

The code below is like 10000 tests and kinda does the trick to trigger the problem.

```c#
public class TestsNumber
{
    public static IEnumerable<int> GetNumbers()
    {
        return Enumerable.Range(1, 10000);
    }

    [TestCaseSource(nameof(GetNumbers))]
    public void TestNumber(int number)
    {
        Console.WriteLine(number);
        Assert.True(true);
    }
}
```

In that case the test run and fail at the same time:

<img width="1039" height="822" alt="image" src="https://github.com/user-attachments/assets/4941690a-ddd5-40c3-a031-3e99f24de9a7" />

**Reply by ricaun** [reply]
@paulbuechner try the version 1.10.2-alpha: https://www.nuget.org/packages/ricaun.RevitTest.TestAdapter/1.10.2-alpha

**Reply by paulbuechner** [reply]
This looks like the issue I'm suffering from. I did try the new version, but to no success, unfortunately.

<img width="799" height="145" alt="grafik" src="https://github.com/user-attachments/assets/4dfb5b6e-e5a5-4cc7-b5d2-c331e8b00d03" />

**Reply by ricaun** [reply]
> This looks like the issue I'm suffering from. I did try the new version, but to no success, unfortunately.

Did you close Revit to make sure the App inside Revit is 1.10.2?

**Reply by paulbuechner** [reply]
Yes, I even restarted my system.

<img width="615" height="121" alt="grafik" src="https://github.com/user-attachments/assets/059d00b9-70e2-4089-965d-9d4dd3674653" />

But again, all Tests pass (will even be marked with a green checkmark in the IDE), but then it somehow doesn't recognize them as finished, waits for the timeout, and then errors.

**Reply by ricaun** [reply]
> Yes, I even restarted my system.
>
> But again, all Tests pass (will even be marked with a green checkmark in the IDE), but then it somehow doesn't recognize them as finished, waits for the timeout, and then errors.

That's annoying and hard to test without a sample that trigger the issue. 
The `Timeout not busy for too long 10 seconds.` happens or the `Timeout 10 minutes.`?
What happen if you run in Visual Studio intend of Rider?
Using `dotnet test` trigger the issue?
If you remove/comment half of the tests the issue still happens?

**Reply by paulbuechner** [reply]
Yes: `Timeout not busy for too long 10 seconds.` is being shown. If I set the `RICAUN_REVITTEST_CONSOLE_TIMEOUT_NOT_BUSY_MAX_SECONDS`, then this value will be shown.

>  What happen if you run in Visual Studio intend of Rider?

The exact same. 

> Using dotnet test trigger the issue?

Same, with the following error msg:

```
Acme.Tests.dll : error TESTERROR: 
      MyTest (< 1ms): ErrorMsg: 
      Trace:
      RevitTest: Timeout not busy for too long 10 seconds.
```

>  If you remove/comment half of the tests the issue still happens?

In the `TestFixture` if select a batch of tests (not all) then it seems to work, but the error raises if a certain Limit (around 20ish) of Tests are run. There is no particular test case triggering it, so it depends on execution time maybe?

> That's annoying and hard to test without a sample that trigger the issue.

If you want you can add me to the projects contributes and I'll try to debug it myself then, but I respect if that's not possible for you or that there is some internal tooling used which I can't run anyways.

**Reply by ricaun** [reply]
> In the `TestFixture` if select a batch of tests (not all) then it seems to work, but the error raises if a certain Limit (around 20ish) of Tests are run. There is no particular test case triggering it, so it depends on execution time maybe?

Execution time should not be a problem, could be something with the response size or something.

> If you want you can add me to the projects contributes and I'll try to debug it myself then, but I respect if that's not possible for you or that there is some internal tooling used which I can't run anyways.

The project is public: https://github.com/ricaun-io/ricaun.RevitTest/tree/develop
Here is where the timeout not busy code: https://github.com/ricaun-io/ricaun.RevitTest/blob/cda71a47a15581f258356cb7bdf1c9a82eeb6597/ricaun.RevitTest.Console/Revit/Utils/RevitTestUtils.cs#L251C1-L251C132

**Reply by paulbuechner** [reply]
Ok, I'll try debug into the issue using the local Adapter then.

**Reply by ricaun** [reply]
I open a issue about this problem.
https://github.com/ricaun-io/ricaun.RevitTest/issues/83

You could try to enable RevitTest verbose to level 3 this should show more informarion in the warning when running the tests.

**Comment by JoannaSieranskiPS** [comment]
Thank you for your help and quick response.

I think the "Details not available" message comes from the test runner. I'm using Rider with the same configuration mentioned in one of the discussions here. It seems that it cannot access the test output and can only write that message.

Regarding the code, I created a smaller example to reproduce the behavior locally. In the attached files, I use two EventHandlers to get ViewSchedules and their names. I then go over the rows and map them to a TableRecord which is basically a list of CellData elements. These are only wrapper objects around the data that I need later. As for the amount of data, I have 19 ViewSchedules with 5–20 rows in it, but the number of columns may vary from 5 to 80. So there are many cells to go over. 
[testQuery.zip](https://github.com/user-attachments/files/18168393/testQuery.zip)

**Reply by ricaun** [reply]
Why do you have the `EventHandler` looks complicated. You should remove that, and run the method directly.

About the `Rider` you could try to run the `dotnet test your_test_file.dll -v d` to run the test directly using the dotnet to see if that works or not.

**Comment by JoannaSieranskiPS** [comment]
Thank you for your suggestions. I already tried running dotnet from the command line and the result is the same:
```
  Failed TestTableUniqueId
  Stack Trace:
  RevitTest: Timeout not busy for too long 10 seconds.
```
As for the EventHandler, I need to use them in production for some reason. Thus, I also keep using them in the tests. But if this causes the problems, then maybe I should refactor my code so I can test more easily.

**Reply by ricaun** [reply]
The `CellData` and `TableRecord` is missing in the sample.

**Reply by JoannaSieranskiPS** [reply]
Sorry for that. Please find the code below.

`public class TableRecord
{
    private readonly List<CellData> _rowData = new List<CellData>();
    private readonly int _recordNumber;
    private readonly string _parentViewName;
    private readonly long _parentViewId;

    public TableRecord(List<CellData> rowData, int recordNumber, ViewSchedule viewSchedule)
    {
        _rowData.AddRange(rowData);
        _recordNumber = recordNumber;
        _parentViewName = viewSchedule.Name ?? "";
        _parentViewId = viewSchedule.Id.Value;
    }

    public string Get(string columnName)
    {
        if (string.IsNullOrEmpty(columnName))
        {
            throw new InvalidOperationException("Column name wasn't provided to get table row content.");
        }

        var cellDatas = _rowData.Where(cellData => columnName.Equals(cellData.GetColumnName())).ToList();
        if (cellDatas.Count == 0)
        {
            throw new InvalidOperationException(
                $"Failed to find column '{columnName}' in record (table '{_parentViewId}' - '{_parentViewName}').");
        }
        
        return cellDatas[0].CellText;
    }

    public int GetRecordNumber()
    {
        return _recordNumber;
    }

    public List<CellData> GetRowData()
    {
        return new List<CellData>(_rowData);
    }
}`

`public class CellData
{
    public readonly string CellText;
    private readonly ParameterElement? _cellCategory;
    private readonly BuiltInParameter? _cellCategoryBuiltIn;
    private readonly ParameterElement? _cellParam;
    private readonly BuiltInParameter? _cellParamBuiltIn;
	
    public CellData(string? cellText, ParameterElement? cellCategory, ParameterElement? cellParam, long cellCategoryId,
        long cellParameterId)
    {
        CellText = cellText ?? "";
        _cellCategory = cellCategory;
        _cellParam = cellParam;

        if (Enum.IsDefined(typeof(BuiltInParameter), cellCategoryId) &&
            cellCategoryId != (long)BuiltInParameter.INVALID)
        {
            _cellCategoryBuiltIn = (BuiltInParameter)Enum.ToObject(typeof(BuiltInParameter), cellCategoryId);
        }

        if (Enum.IsDefined(typeof(BuiltInParameter), cellParameterId) &&
            cellParameterId != (long)BuiltInParameter.INVALID)
        {
            _cellParamBuiltIn = (BuiltInParameter)Enum.ToObject(typeof(BuiltInParameter), cellParameterId);
        }
    }

    public string? GetColumnName()
    {
        if (_cellCategory != null || _cellParam != null)
        {
            return _cellCategory?.Name ?? _cellParam?.Name;
        }

        if (_cellCategoryBuiltIn != null)
        {
            return LabelUtils.GetLabelFor((BuiltInParameter)_cellCategoryBuiltIn);
        }

        if (_cellParamBuiltIn != null)
        {
            return LabelUtils.GetLabelFor((BuiltInParameter)_cellParamBuiltIn);
        }

        return null;
    }

    public string? GetCellType()
    {
        var definition = _cellCategory?.GetDefinition() ?? _cellParam?.GetDefinition();
        return definition?.ParameterGroup.ToString();
    }
}`

**Reply by ricaun** [reply]
Great! Can I create a repo with the code? MIT or something? 

You could create one as well,I create a base template for the RevitTest in the https://www.nuget.org/packages/ricaun.Revit.Templates

After install should have the `ricaun RevitTest Project`, I suppose in Rider should show as well.
![image](https://github.com/user-attachments/assets/dd4a263f-92db-43ec-a9fd-2b64e56fa187)

**Reply by ricaun** [reply]
I did some test but was not able to trigger the `Timeout not busy for too long 10 seconds` using your code.
That timeout should never happen in a normal run, what version of the RevitTest are you using?

**Reply by JoannaSieranskiPS** [reply]
If it helps to find the issue, you may put the code on the Revit Templates repo.

I started the tests with version 1.6.0 and updated to 1.7.0 and 1.7.1 recently. However, if you can't reproduce the behavior, then maybe there is something wrong with my data. Unfortunately, I cannot share the Revit project.

**Reply by ricaun** [reply]
The file should not interfere I guess, did you tested with some other files? Like the Revit sample?

**Reply by JoannaSieranskiPS** [reply]
I just copied the files to the Revit Sample project and strangely the test works there. No timeouts occur. I'll check my project setup, and compare it to the sample. Thank you for investigating the issue.
